lb4 CLI create controller bug - Filter object missing an argument

Run `lb4 controller` then run `npm start` it will complain argument of type 'Filter<AnyObject> undefined' see detail on below(Log).
Auto generated code by CLI is missing an argument on Filter object.
_**See comparison on below:**_
**Generated code:**
`async find(
    @param.query.object('filter', getFilterSchemaFor(Todo)) filter?: Filter,
  ): Promise<Todo[]> {
    return await this.todoRepository.find(filter);
  }`
**Correct version:**
`async find(
    @param.query.object('filter', getFilterSchemaFor(Todo)) filter?: Filter<Todo>,
  ): Promise<Todo[]> {
    return await this.todoRepository.find(filter);
  }`

**Environment:**
npm v.6.9.0
@loopback/cli version: 1.14.1

**Log:**
```
✘-INT ~/Desktop/todo-list
22:14 $ lb4 controller
? Controller class name: todo
? What kind of controller would you like to generate? REST Controller with CRUD functions
? What is the name of the model to use with this CRUD repository? Todo
? What is the name of your CRUD repository? TodoRepository
? What is the type of your ID? number
? What is the base HTTP path name of the CRUD operations? /todos
   create src/controllers/todo.controller.ts
   update src/controllers/index.ts

Controller todo was created in src/controllers/

✔ ~/Desktop/todo-list
22:15 $ npm start

> todo-list@1.0.0 prestart /Users/******/Desktop/todo-list
> npm run build


> todo-list@1.0.0 build /Users/******/Desktop/todo-list
> lb-tsc es2017 --outDir dist

src/controllers/todo.controller.ts:69:43 - error TS2345: Argument of type 'Filter<AnyObject> | undefined' is not assignable to parameter of type 'Filter<Todo> | undefined'.
  Type 'Filter<AnyObject>' is not assignable to type 'Filter<Todo>'.
    Type 'AnyObject' is missing the following properties from type 'Todo': title, getId, getIdObject, toJSON, toObject

69     return await this.todoRepository.find(filter);
                                             ~~~~~~


Found 1 error.

npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! todo-list@1.0.0 build: `lb-tsc es2017 --outDir dist`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the todo-list@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/******/.npm/_logs/2019-06-02T02_15_08_550Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! todo-list@1.0.0 prestart: `npm run build`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the todo-list@1.0.0 prestart script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/******/.npm/_logs/2019-06-02T02_15_08_567Z-debug.log
```

--- a/docs/site/Controller-generator.md
+++ b/docs/site/Controller-generator.md
@@ -136,7 +136,7 @@ export class TodoController {
     },
   })
   async count(
-    @param.query.object('where', getWhereSchemaFor(Todo)) where?: Where,
+    @param.query.object('where', getWhereSchemaFor(Todo)) where?: Where<Todo>,
   ): Promise<number> {
     return await this.todoRepository.count(where);
   }
@@ -154,7 +154,8 @@ export class TodoController {
     },
   })
   async find(
-    @param.query.object('filter', getFilterSchemaFor(Todo)) filter?: Filter,
+    @param.query.object('filter', getFilterSchemaFor(Todo))
+    filter?: Filter<Todo>,
   ): Promise<Todo[]> {
     return await this.todoRepository.find(filter);
   }
@@ -169,7 +170,7 @@ export class TodoController {
   })
   async updateAll(
     @requestBody() data: Todo,
-    @param.query.object('where', getWhereSchemaFor(Todo)) where?: Where,
+    @param.query.object('where', getWhereSchemaFor(Todo)) where?: Where<Todo>,
   ): Promise<number> {
     return await this.todoRepository.updateAll(data, where);
   }

--- a/docs/site/Repositories.md
+++ b/docs/site/Repositories.md
@@ -23,7 +23,7 @@ operations. These `Repository` implementations leverage `Model` definition and
 interface Repository<T extends Model> {}
 
 interface CustomerRepository extends Repository<Customer> {
-  find(filter?: Filter, options?: Options): Promise<Customer[]>;
+  find(filter?: Filter<Customer>, options?: Options): Promise<Customer[]>;
   findByEmail(email: string): Promise<Customer>;
   // ...
 }
@@ -492,10 +492,10 @@ retrieve all the rows that match a particular filter for a model instance.
 
 ```ts
 public find(
-  modelClass: Class<Entity>,
-  filter: Filter,
+  modelClass: Class<Account>,
+  filter: Filter<Account>,
   options: Options
-): Promise<EntityData[]> {
+): Promise<Account[]> {
   let self = this;
   let sqlStmt = "SELECT * FROM " + modelClass.name;
   if (filter.where) {

--- a/docs/site/tutorials/todo-list/todo-list-tutorial-controller.md
+++ b/docs/site/tutorials/todo-list/todo-list-tutorial-controller.md
@@ -154,7 +154,7 @@ export class TodoListTodoController {
   })
   async find(
     @param.path.number('id') id: number,
-    @param.query.object('filter') filter?: Filter,
+    @param.query.object('filter') filter?: Filter<Todo>,
   ): Promise<Todo[]> {
     return await this.todoListRepo.todos(id).find(filter);
   }
@@ -170,7 +170,7 @@ export class TodoListTodoController {
   async patch(
     @param.path.number('id') id: number,
     @requestBody() todo: Partial<Todo>,
-    @param.query.object('where', getWhereSchemaFor(Todo)) where?: Where,
+    @param.query.object('where', getWhereSchemaFor(Todo)) where?: Where<Todo>,
   ): Promise<Count> {
     return await this.todoListRepo.todos(id).patch(todo, where);
   }
@@ -185,7 +185,7 @@ export class TodoListTodoController {
   })
   async delete(
     @param.path.number('id') id: number,
-    @param.query.object('where', getWhereSchemaFor(Todo)) where?: Where,
+    @param.query.object('where', getWhereSchemaFor(Todo)) where?: Where<Todo>,
   ): Promise<Count> {
     return await this.todoListRepo.todos(id).delete(where);
   }

--- a/packages/cli/generators/controller/templates/src/controllers/controller-rest-template.ts.ejs
+++ b/packages/cli/generators/controller/templates/src/controllers/controller-rest-template.ts.ejs
@@ -46,7 +46,7 @@ export class <%= className %>Controller {
     },
   })
   async count(
-    @param.query.object('where', getWhereSchemaFor(<%= modelName %>)) where?: Where,
+    @param.query.object('where', getWhereSchemaFor(<%= modelName %>)) where?: Where<<%= modelName %>>,
   ): Promise<Count> {
     return await this.<%= repositoryNameCamel %>.count(where);
   }
@@ -64,7 +64,7 @@ export class <%= className %>Controller {
     },
   })
   async find(
-    @param.query.object('filter', getFilterSchemaFor(<%= modelName %>)) filter?: Filter,
+    @param.query.object('filter', getFilterSchemaFor(<%= modelName %>)) filter?: Filter<<%= modelName %>>,
   ): Promise<<%= modelName %>[]> {
     return await this.<%= repositoryNameCamel %>.find(filter);
   }
@@ -79,7 +79,7 @@ export class <%= className %>Controller {
   })
   async updateAll(
     @requestBody() <%= modelVariableName %>: <%= modelName %>,
-    @param.query.object('where', getWhereSchemaFor(<%= modelName %>)) where?: Where,
+    @param.query.object('where', getWhereSchemaFor(<%= modelName %>)) where?: Where<<%= modelName %>>,
   ): Promise<Count> {
     return await this.<%= repositoryNameCamel %>.updateAll(<%= modelVariableName %>, where);
   }

--- a/packages/cli/generators/relation/templates/controller-relation-template-has-many.ts.ejs
+++ b/packages/cli/generators/relation/templates/controller-relation-template-has-many.ts.ejs
@@ -39,7 +39,7 @@ export class <%= controllerClassName %> {
   })
   async find(
     @param.path.<%= sourceModelPrimaryKeyType %>('id') id: <%= sourceModelPrimaryKeyType %>,
-    @param.query.object('filter') filter?: Filter,
+    @param.query.object('filter') filter?: Filter<<%= targetModelClassName %>>,
   ): Promise<<%= targetModelClassName %>[]> {
     return await this.<%= paramSourceRepository %>.<%= relationPropertyName %>(id).find(filter);
   }
@@ -70,7 +70,7 @@ export class <%= controllerClassName %> {
   async patch(
     @param.path.<%= sourceModelPrimaryKeyType %>('id') id: <%= sourceModelPrimaryKeyType %>,
     @requestBody() <%= targetModelRequestBody %>: Partial<<%= targetModelClassName %>>,
-    @param.query.object('where', getWhereSchemaFor(<%= targetModelClassName %>)) where?: Where,
+    @param.query.object('where', getWhereSchemaFor(<%= targetModelClassName %>)) where?: Where<<%= targetModelClassName %>>,
   ): Promise<Count> {
     return await this.<%= paramSourceRepository %>.<%= relationPropertyName %>(id).patch(<%= targetModelRequestBody %>, where);
   }
@@ -85,7 +85,7 @@ export class <%= controllerClassName %> {
   })
   async delete(
     @param.path.<%= sourceModelPrimaryKeyType %>('id') id: <%= sourceModelPrimaryKeyType %>,
-    @param.query.object('where', getWhereSchemaFor(<%= targetModelClassName %>)) where?: Where,
+    @param.query.object('where', getWhereSchemaFor(<%= targetModelClassName %>)) where?: Where<<%= targetModelClassName %>>,
   ): Promise<Count> {
     return await this.<%= paramSourceRepository %>.<%= relationPropertyName %>(id).delete(where);
   }

--- a/packages/cli/test/integration/generators/controller.integration.js
+++ b/packages/cli/test/integration/generators/controller.integration.js
@@ -261,7 +261,7 @@ function checkRestCrudContents() {
     /'200': {/,
     /description: 'ProductReview model count'/,
     /content: {'application\/json': {schema: CountSchema}},\s{1,}},\s{1,}},\s{1,}}\)/,
-    /async count\(\s+\@param\.query\.object\('where', getWhereSchemaFor\(ProductReview\)\) where\?: Where(|,\s+)\)/,
+    /async count\(\s+\@param\.query\.object\('where', getWhereSchemaFor\(ProductReview\)\) where\?: Where<ProductReview>(|,\s+)\)/,
   ];
   getCountRegEx.forEach(regex => {
     assert.fileContent(expectedFile, regex);
@@ -274,7 +274,7 @@ function checkRestCrudContents() {
     /'200': {/,
     /description: 'Array of ProductReview model instances'/,
     /content: {'application\/json': {schema: {'x-ts-type': ProductReview}}},\s{1,}},\s{1,}},\s{1,}}\)/,
-    /async find\(\s*\@param\.query\.object\('filter', getFilterSchemaFor\(ProductReview\)\) filter\?: Filter(|,\s+)\)/,
+    /async find\(\s*\@param\.query\.object\('filter', getFilterSchemaFor\(ProductReview\)\) filter\?: Filter<ProductReview>(|,\s+)\)/,
   ];
   getFindRegEx.forEach(regex => {
     assert.fileContent(expectedFile, regex);
@@ -287,7 +287,7 @@ function checkRestCrudContents() {
     /'200': {/,
     /description: 'ProductReview PATCH success count'/,
     /content: {'application\/json': {schema: CountSchema}},\s{1,}},\s{1,}},\s{1,}}\)/,
-    /async updateAll\(\s{1,}\@requestBody\(\) productReview: ProductReview,\s{1,} @param\.query\.object\('where', getWhereSchemaFor\(ProductReview\)\) where\?: Where(|,\s+)\)/,
+    /async updateAll\(\s{1,}\@requestBody\(\) productReview: ProductReview,\s{1,} @param\.query\.object\('where', getWhereSchemaFor\(ProductReview\)\) where\?: Where<ProductReview>(|,\s+)\)/,
   ];
   patchUpdateAllRegEx.forEach(regex => {
     assert.fileContent(expectedFile, regex);

--- a/packages/cli/test/integration/generators/hasmany.relation.integration.js
+++ b/packages/cli/test/integration/generators/hasmany.relation.integration.js
@@ -386,7 +386,7 @@ context('check if the controller file created ', () => {
           /schema: { type: 'array', items: { 'x-ts-type': Order } },/,
           /},\n . {6}},\n . {4}},\n . {2}},\n {2}}\)\n/,
           /async find\(\n . {2}\@param.path.number\('id'\) id: number,\n/,
-          /\@param.query.object\('filter'\) filter\?: Filter,\n/,
+          /\@param.query.object\('filter'\) filter\?: Filter<Order>,\n/,
           /\)\: Promise<Order\[]> {\n/,
           /return await this\.customerRepository\.orders\(id\)\.find\(filter\);\n {2}}\n/,
         ];
@@ -397,7 +397,7 @@ context('check if the controller file created ', () => {
           /schema: { type: 'array', items: { 'x-ts-type': OrderClass } },/,
           /},\n . {6}},\n . {4}},\n . {2}},\n {2}}\)\n/,
           /async find\(\n . {2}\@param.path.number\('id'\) id: number,\n/,
-          /\@param.query.object\('filter'\) filter\?: Filter,\n/,
+          /\@param.query.object\('filter'\) filter\?: Filter<OrderClass>,\n/,
           /\)\: Promise<OrderClass\[]> {\n/,
           /return await this\.customerClassRepository\.orderClasses\(id\)\.find\(filter\);\n {2}}\n/,
         ];
@@ -408,7 +408,7 @@ context('check if the controller file created ', () => {
           /schema: { type: 'array', items: { 'x-ts-type': OrderClassType } },/,
           /},\n . {6}},\n . {4}},\n . {2}},\n {2}}\)\n/,
           /async find\(\n . {2}\@param.path.number\('id'\) id: number,\n/,
-          /\@param.query.object\('filter'\) filter\?: Filter,\n/,
+          /\@param.query.object\('filter'\) filter\?: Filter<OrderClassType>,\n/,
           /\)\: Promise<OrderClassType\[]> {\n/,
           /return await this\.customerClassTypeRepository\.orderClassTypes\(id\).find\(filter\);\n {2}}\n/,
         ];
@@ -495,7 +495,7 @@ context('check if the controller file created ', () => {
           /content: { 'application\/json': { schema: CountSchema } },\n/,
           /},\n {4}},\n {2}}\)\n {2}async patch\(\n/,
           /\@param\.path\.number\('id'\) id: number,\n {4}\@requestBody\(\) order: Partial<Order>,\n/,
-          /\@param\.query\.object\('where', getWhereSchemaFor\(Order\)\) where\?: Where,\n/,
+          /\@param\.query\.object\('where', getWhereSchemaFor\(Order\)\) where\?: Where<Order>,\n/,
           /\): Promise<Count> {\n/,
           /return await this\.customerRepository\.orders\(id\).patch\(order, where\);\n {2}}\n/,
         ];
@@ -506,7 +506,7 @@ context('check if the controller file created ', () => {
           /content: { 'application\/json': { schema: CountSchema } },\n/,
           /},\n {4}},\n {2}}\)\n {2}async patch\(\n/,
           /\@param\.path\.number\('id'\) id: number,\n {4}\@requestBody\(\) orderClass: Partial<OrderClass>,\n/,
-          /\@param\.query\.object\('where', getWhereSchemaFor\(OrderClass\)\) where\?: Where,\n/,
+          /\@param\.query\.object\('where', getWhereSchemaFor\(OrderClass\)\) where\?: Where<OrderClass>,\n/,
           /\): Promise<Count> {\n/,
           /return await this\.customerClassRepository\.orderClasses\(id\)\.patch\(orderClass, where\);\n {2}}\n/,
         ];
@@ -517,7 +517,7 @@ context('check if the controller file created ', () => {
           /content: { 'application\/json': { schema: CountSchema } },\n/,
           /},\n {4}},\n {2}}\)\n {2}async patch\(\n/,
           /\@param\.path\.number\('id'\) id: number,\n {4}\@requestBody\(\) orderClassType: Partial<OrderClassType>,\n/,
-          /\@param\.query\.object\('where', getWhereSchemaFor\(OrderClassType\)\) where\?: Where,\n/,
+          /\@param\.query\.object\('where', getWhereSchemaFor\(OrderClassType\)\) where\?: Where<OrderClassType>,\n/,
           /\): Promise<Count> {\n/,
           /return await this\.customerClassTypeRepository\.orderClassTypes\(id\).patch\(orderClassType, where\);\n {2}}\n/,
         ];
@@ -553,7 +553,7 @@ context('check if the controller file created ', () => {
           /content: { 'application\/json': { schema: CountSchema } },\n/,
           /},\n {4}},\n {2}}\)\n {2}async delete\(\n/,
           /\@param\.path\.number\('id'\) id: number,\n /,
-          /\@param\.query\.object\('where', getWhereSchemaFor\(Order\)\) where\?: Where,\n/,
+          /\@param\.query\.object\('where', getWhereSchemaFor\(Order\)\) where\?: Where<Order>,\n/,
           /\): Promise<Count> {\n/,
           /return await this\.customerRepository\.orders\(id\)\.delete\(where\);\n {2}}\n}\n/,
         ];
@@ -564,7 +564,7 @@ context('check if the controller file created ', () => {
           /content: { 'application\/json': { schema: CountSchema } },\n/,
           /},\n {4}},\n {2}}\)\n {2}async delete\(\n/,
           /\@param\.path\.number\('id'\) id: number,\n /,
-          /\@param\.query\.object\('where', getWhereSchemaFor\(OrderClass\)\) where\?: Where,\n/,
+          /\@param\.query\.object\('where', getWhereSchemaFor\(OrderClass\)\) where\?: Where<OrderClass>,\n/,
           /\): Promise<Count> {\n/,
           /return await this\.customerClassRepository\.orderClasses\(id\)\.delete\(where\);\n {2}}\n}\n/,
         ];
@@ -575,7 +575,7 @@ context('check if the controller file created ', () => {
           /content: { 'application\/json': { schema: CountSchema } },\n/,
           /},\n {4}},\n {2}}\)\n {2}async delete\(\n/,
           /\@param\.path\.number\('id'\) id: number,\n /,
-          /\@param\.query\.object\('where', getWhereSchemaFor\(OrderClassType\)\) where\?: Where,\n/,
+          /\@param\.query\.object\('where', getWhereSchemaFor\(OrderClassType\)\) where\?: Where<OrderClassType>,\n/,
           /\): Promise<Count> {\n/,
           /return await this\.customerClassTypeRepository\.orderClassTypes\(id\)\.delete\(where\);\n {2}}\n}\n/,
         ];

